src/subscriptions/__tests__/subscriptionsReducer-test.js

```diff
@@ -1,131 +1,69 @@
+/* @flow strict-local */
+
 import deepFreeze from 'deep-freeze';
 
 import { EventTypes } from '../../api/eventTypes';
-import { REALM_INIT, EVENT_SUBSCRIPTION, ACCOUNT_SWITCH, EVENT } from '../../actionConstants';
+import { EVENT_SUBSCRIPTION, ACCOUNT_SWITCH, EVENT } from '../../actionConstants';
 import subscriptionsReducer from '../subscriptionsReducer';
+import * as eg from '../../__tests__/lib/exampleData';
 
 describe('subscriptionsReducer', () => {
+  const stream1 = eg.makeStream({ name: 'stream1', description: 'my first stream' });
+  const sub1 = eg.makeSubscription({ stream: stream1 });
+
+  const stream2 = eg.makeStream({ name: 'stream2', description: 'my second stream' });
+  const sub2 = eg.makeSubscription({ stream: stream2 });
+
+  const stream3 = eg.makeStream({ name: 'stream3', description: 'my third stream' });
+  const sub3 = eg.makeSubscription({ stream: stream3 });
+
   describe('REALM_INIT', () => {
     test('when `subscriptions` data is provided init state with it', () => {
       const initialState = deepFreeze([]);
+
       const action = deepFreeze({
-        type: REALM_INIT,
+        ...eg.action.realm_init,
         data: {
-          subscriptions: [
-            {
-              name: 'some stream',
-              stream_id: 1,
-            },
-          ],
+          ...eg.action.realm_init.data,
+          subscriptions: [sub1],
         },
       });
 
       const actualState = subscriptionsReducer(initialState, action);
 
-      expect(actualState).toEqual([
-        {
-          name: 'some stream',
-          stream_id: 1,
-        },
-      ]);
-    });
-
-    test('when no `subscriptions` data is given reset state', () => {
```

Comment:
This sure makes sense.

In this commit, my reason for removing these two tests (and, whether or not it's right, I should have written it down) is simply that, in each of them, it's impossible to represent their test input without defying our type definitions. So, in live code, the same bad input will be caught by the type checker, making any Jest tests that use that input redundant. What do you think; is this wrong, in general?

On a closer look, something seems weird with the bit of the reducer this test is for. I would expect Flow to complain after the `||`, where `action.data.subscriptions` is falsy —

```
    case REALM_INIT:
      return action.data.subscriptions
        || initialState; // should see a Flow error, but there is none
```

— but it doesn't. An inspection of RealmInitAction, following the aliases all the way down, shows that `data.subscriptions` is not an optional property, nor is it unioned with `void`.

Aha, and, indeed, a type-checked version of this test passes Flow with no errors, which I _definitely_ did not expect, following my reading of RealmInitAction:

```
  test('when no `subscriptions` data is given reset state', () => {
    const initialState = deepFreeze([sub1]);
    const action = deepFreeze({
      ...eg.action.realm_init,
      data: {
        ...eg.action.realm_init.data,
        subscriptions: undefined, // no problems here! Why?
      },
    });
    const expectedState = [];

    const actualState = subscriptionsReducer(initialState, action);

    expect(actualState).toEqual(expectedState);
  });
```

So, I guess, other questions:

1) Am I reading our type definitions correctly, that `action.data.subscriptions` must never be missing, and
2) If so, how can we get away with expressing it as missing, anywhere? Maybe there's a Flow bug with spread types affecting the InitialData type? And, I guess,
3) Assuming I'm reading the type definitions correctly, and that they are our best knowledge of what will come from the server, do we even need to defend against `data.subscriptions` being missing, in our live code?

Reply:
Ah OK, no problem! Thanks for confirming this is a known Flow bug; glad there's an eventual solution, and good to be aware of the problem in the meantime.

Reply:
> (I do think a test with the former would be valuable, and that it would be better to turn this test into that test than to eliminate it entirely; but that's a very different thing than what I said earlier. 😓)

How about point 3, [above](https://github.com/zulip/zulip-mobile/pull/4046#discussion_r412543382)? Is this bit of live code in `subscriptionsReducer` reasonable, in a world where we don't expect `.subscriptions` to be missing?

```
    case REALM_INIT:
      return action.data.subscriptions || initialState;
```

An empty array is truthy, so I'm not sure what work `|| initialState;` is meant to accomplish. Following from that, I'm not sure how to proceed with modifying the test as you've recommended.

Reply:
What do you think about removing ` || initialState;`? Seems kind of scary, actually.

Reply:
> What do you think about removing ` || initialState;`? Seems kind of scary, actually, but its presence tells me something's not quite right with the status quo (whether here, or in our types, I'm not sure).

Looks like that originates in 462215c26fa5acc9092ab3789c672b5eb59f0e72. Following links, that was from #2878 which fixed #2859. The latter issue was that, in a *different* reducer, we needed exactly this because the data in question was about a recent feature that some servers didn't support and so the property (namely `realm_user_groups`) would be missing.

For `subscriptions` the same reasoning doesn't apply, I think :slightly_smiling_face: -- it'd be sensible to take that condition out.

Reply:
One small bug that story brings out is that we don't have `realm_user_groups` marked as optional in our API types, and we should:
```
export type InitialDataRealmUserGroups = {|
  realm_user_groups: UserGroup[],
|};
```
along with I guess a comment that it was new in 1.8.0 (or whatever).

There's lots of other stuff in that `initialDataTypes` module that needs a review for similar questions; the bulk of it is entirely uncommented. That's well afield from this PR, though.

Reply:
Alas, your point 3 starts out with 
> Assuming I'm reading the type definitions correctly, and that they are our best knowledge of what will come from the server,

The first assumption is probably correct; the second... less so. I wouldn't trust that the server has never handed out `null` or similar there, at least not without an archaeological expedition.

[...]

Having done one of those, I think it hasn't. As of at least zulip/zulip@7a930afa07be2e2676c277a6a2e1b04b22d60410 (2017-02-20), and probably much earlier, it should always send at least `[]` for `subscriptions` so long as `subscription` is specified as a desired event type in the `/register` call. (I'm pretty sure that's true both at that commit and in today's code; but there *are* a lot of intermediate changes that I haven't checked fully, so I'm less sure it's been true the entire time.)

As Greg notes, the real reason can be seen in #2878, and a search of our codebase for the string `|| initialState` shows that many of our reducers keep the same pattern even today.

Reply:
> Having done one of those, I think it hasn't. As of at least [zulip/zulip@7a930af](https://github.com/zulip/zulip/commit/7a930afa07be2e2676c277a6a2e1b04b22d60410) (2017-02-20), and probably much earlier, it should always send at least `[]` for `subscriptions` so long as `subscription` is specified as a desired event type in the `/register` call.

Cool! Please add a comment with that result, so we can memoize that archaeological work :slightly_smiling_face: 


> (I'm pretty sure that's true both at that commit and in today's code; but there _are_ a lot of intermediate changes that I haven't checked fully, so I'm less sure it's been true the entire time.)

In general when a piece of the API was the same in an old server version as in a later one, I'm content to just record that fact and then act on the assumption that it didn't in between change to something else and change back.

I'm not aware that an A -> B -> A change like that has ever happened in the Zulip API; and if such a case ever does arise, we can learn about it through Sentry or a bug report. And yeah, it'd often be quite a lot more work to check everything in between.

Reply:
> Please add a comment with that result, so we can memoize that archaeological work 🙂

I was just doing that as we speak. Thanks, Ray! 🙂

— as in, I was just putting it in a commit message. Would you say we should have a code comment as well?

I'm seeing a subtle lesson coming from this.

On the one hand, it was likely skepticism and caution, in healthy doses, that motivated adding the `|| initialState` check, and in several places — skepticism and caution are good motivators that drive the best design / code choices.

...but `|| initialState` is dead code, and dead code should be removed (particularly when it makes our types look funny), and it's *not at all trivial* to prove that it's dead. You can't just make it go away with some Boolean algebra, for example. While the check was well-intentioned, its net effect has been negative, causing us all to go and do these investigations.

So, I'll keep this case in mind when I try to add "crunchy shell" checks, in future, which might turn out to be dead code that takes work to uproot. I sincerely hope that https://github.com/zulip/zulip-mobile/pull/3966/commits/923e2b447c434d63f76d95cb1d730ca496785119 doesn't come back to bite us in this way; some small amount of context is in the latter half of https://github.com/zulip/zulip-mobile/pull/3966#issuecomment-611795320.

Reply:
> — as in, I was just putting it in a commit message. Would you say we should have a code comment as well?

Yeah, on the type definition. See examples in `initialDataTypes.js` (for the few types there that are commented...) and in `modelTypes.js`. That's a handy place to later look up what we know.



> So, I'll keep this case in mind when I try to add "crunchy shell" checks, in future, which might turn out to be dead code that takes work to uproot.

Yeah, so I would describe this `|| initialState` check not as part of a "crunchy shell" but rather as the kind of thing we end up with in the middle of the app [as a result of](https://github.com/zulip/zulip-mobile/blob/master/docs/architecture/crunchy-shell.md#more-needed) a soft, non-crunchy shell:
> In our JS code and the data it handles from HTTP requests it makes to the Zulip server, there are a lot of spots that would benefit from a crunchy shell -- a lot of spots where the code in the middle of the app either has random checks and conditionals scattered through, or bugs from the lack of those. 

And yep, this is exactly the kind of thing that can then happen with such code :)

Reply:
> Yeah, so I would describe this `|| initialState` check not as part of a "crunchy shell" but rather as the kind of thing we end up with in the middle of the app

It sounds like the same would apply to 923e2b4, no? From rereading the "crunchy shell" doc, it sounds like the ultimate goal is to get things thoroughly vetted in a layer before the data even reaches our Redux reducers. If doing these checks in a reducer is considered "in the middle of the app", 923e2b4 is thoroughly guilty.

Reply:
> One small bug that story brings out is that we don't have `realm_user_groups` marked as optional in our API types, and we should:
> 
> ```
> export type InitialDataRealmUserGroups = {|
>   realm_user_groups: UserGroup[],
> |};
> ```
> 
> along with I guess a comment that it was new in 1.8.0 (or whatever).
> 
> There's lots of other stuff in that `initialDataTypes` module that needs a review for similar questions; the bulk of it is entirely uncommented. That's well afield from this PR, though.

Opened as https://github.com/zulip/zulip-mobile/pull/4073.

Reply:
> It sounds like the same would apply to [923e2b4](https://github.com/zulip/zulip-mobile/commit/923e2b447c434d63f76d95cb1d730ca496785119), no? From rereading the "crunchy shell" doc, it sounds like the ultimate goal is to get things thoroughly vetted in a layer before the data even reaches our Redux reducers. If doing these checks in a reducer is considered to be "in the middle of the app", then [923e2b4](https://github.com/zulip/zulip-mobile/commit/923e2b447c434d63f76d95cb1d730ca496785119) is thoroughly guilty.

Yeah, it's true that isn't the ideal spot in the app for the check to be happening. It still feels better to me in the case of [923e2b4](https://github.com/zulip/zulip-mobile/commit/923e2b447c434d63f76d95cb1d730ca496785119) than here. Maybe one reason is that for [923e2b4](https://github.com/zulip/zulip-mobile/commit/923e2b447c434d63f76d95cb1d730ca496785119) we'd looked and determined that the guarantee we needed wasn't something clearly specified in the API; whereas here it is, and so adding the check gives me the sense that the author just hadn't bothered to determine what the API really was and base the code on that.